### ISSUE TITLE
inpainting implement different from SD-webui

### ISSUE THREAD

**Issue by AmberCheng** [huggingface/diffusers#3514]
hi, recently I tried the diffuser inpaintpipeline and the SD-webui img2img inpaint, I found the result is different. As for the diffuser have change the part I have masked for unchanging. I wonder what the differences from these two method?

**Comment by AmberCheng** [comment]
> Hey @AmberCheng,
> 
> Could you add some more links here? It's impossible to find out what exactly might be the difference from just reading this. Thanks!

I'm sorry for incomplete information. I have used the inpaintpipeline to generate the background as I unmasked. The img is like this: 
 
![test_goods](https://github.com/huggingface/diffusers/assets/18375990/92a8e0d9-b5ea-4136-bb17-eb7135ddbdc6)

result is like this:
![res_3](https://github.com/huggingface/diffusers/assets/18375990/4e728484-179d-4bed-a7b1-aa8be30c3d3f)

I found that the object that I want to use mask to keep had been changed.

Also, I attach the code I used:

from diffusers import StableDiffusionInpaintPipeline
import imageio
import numpy as np
from PIL import Image
import torch
import cv2

rgba_image = imageio.imread('test_goods.png')
org_size = (rgba_image.shape[1],rgba_image.shape[0])
rgba_image = cv2.resize(rgba_image, (512,512))

image = rgba_image[...,:3]
image = Image.fromarray(image)

mask_image = 255 - rgba_image[...,3:]
mask_image = mask_image.repeat(3,axis=2)
mask_image = Image.fromarray(mask_image)

pipe = StableDiffusionInpaintPipeline.from_pretrained(
    # "runwayml/stable-diffusion-inpainting",
    "stabilityai/stable-diffusion-2-inpainting",
    # revision="fp16",
    # torch_dtype=torch.float16,
)
pipe = pipe.to("cuda")
prompt = 'high resolution, a perfume with flowers,soft light'
#image and mask_image should be PIL images.
#The mask structure is white for inpainting and black for keeping as is
images = pipe(prompt=prompt, image=image, mask_image=mask_image,num_images_per_prompt=8).images

for idx, image in enumerate(images): 
    image = image.resize(org_size, Image.ANTIALIAS)
    image.save("./res_%s.png" % idx)

**Comment by rupertmenneer** [comment]
Hi @AmberCheng I am sure that @patrickvonplaten will know much more than me, but as far as I know this makes sense given the process of the `StableDiffusionInpaintPipeline`:

1. Encodes your masked image into latent space through the autoencoder's encoder (this compresses the image)
2. Runs the inpainting diffusion process in latent space, as it is much more efficient to do so
3. Finally once the diffusion process has finished, the autoencoder's decoder coverts the in-painted latents back into image space.

I don't believe there is any guarantee that non-masked pixels end up identical in the final output due to image being passed through the autoencoder (compression -> decompression). However, this guarantee could be easily added as a post processing step, e.g. a boolean to the pipeline called `return_identical_unmasked` - essentially this would iterate over the output image and swap out the real input pixels with output pixels in unmasked regions ensuring you have that guarantee. In practice I am not sure whether this would make outputs 'less smooth' or create jaggies so something worth investigating first. By the looks of things it might make sense to add into the `VAEImageProcessor` as an option. What are your thoughts @patrickvonplaten ? Happy to add an issue / pr if that sounds like something you'd like to include as an option, or if this is already a thing on other pipelines, etc.

**Comment by Markus-Pobitzer** [comment]
Dear @AmberCheng, I also think this is expected behavior as @rupertmenneer mentioned.

I created a small tutorial notebook (you can also run it on colab, but make sure to select a GPU runtime) here: [https://github.com/Markus-Pobitzer/Inpainting-Tutorial/tree/main](https://github.com/Markus-Pobitzer/Inpainting-Tutorial/tree/main). Open notebook and scroll to "Keep original image in unmasked area" section.

Basically what you want is a function like this:
```
from PIL import Image
from PIL import ImageFilter

# This function blends the inpainted image with the original image using a mask
# blur is used to better blend the two images together
def blend_image(inpainted, original, mask, blur=3,):
    mask = mask.convert("L")
    # Apply blur
    mask = mask.filter(ImageFilter.GaussianBlur(blur))
    # Blend images together
    return Image.composite(inpainted.convert('RGBA'), original.convert('RGBA'), mask).convert('RGBA')
```

As far as I have seen, SD webui does something similar
